ArduCopter/mode_autorotate.cpp

```diff
@@ -0,0 +1,417 @@
+                float pilot_spd_dn = -get_pilot_speed_dn();
+                float pilot_spd_up = g.pilot_speed_up;
+
+                //set speed limit
+                pos_control->set_max_speed_z(curr_vel_z, pilot_spd_up);
+
+                float pilot_des_v_z = get_pilot_desired_climb_rate(channel_throttle->get_control_in());
+                pilot_des_v_z = constrain_float(pilot_des_v_z, pilot_spd_dn, pilot_spd_up);
+
+                //calculate target climb rate adjustment to transition from bail out decent speed to requested climb rate on stick.
+                _target_climb_rate_adjust = (curr_vel_z - pilot_des_v_z)/(_param_bail_time - BAILOUT_RAMP_TIME); //accounting for 0.5s motor spool time
+
+                //calculate pitch target adjustment rate to return to level
+                _target_pitch_adjust = _pitch_target/(_param_bail_time - BAILOUT_RAMP_TIME); //accounting for 0.5s motor spool time_param_bail_time;
+
+                //set acceleration limit
+                pos_control->set_max_accel_z(abs(_target_climb_rate_adjust));
+
+                motors->set_desired_spool_state(AP_Motors::DesiredSpoolState::THROTTLE_UNLIMITED);
+
+                _flags.bail_out_initial = 0;
+            }
+
+        if (bail_time_remain >= BAILOUT_RAMP_TIME) {
+            //update desired vertical speed and pitch target after the bailout ramp timer has completed
+            _desired_v_z -= _target_climb_rate_adjust*G_Dt;
+            _pitch_target -= _target_pitch_adjust*G_Dt;
+        }
+        //set position controller
+        pos_control->set_alt_target_from_climb_rate(_desired_v_z, G_Dt, false);
+
+        //update controllers
+        pos_control->update_z_controller();
+
+        //update bail out timer
+        bail_time_remain -= G_Dt;
+        
+        if (bail_time_remain <= 0.0f) {
+            //bail out timer complete.  Change flight mode.
+            set_mode(copter.prev_control_mode, MODE_REASON_AUTOROTATION_BAILOUT);
+        }
+
+        break;
+        }
+    }
+
+
+    switch (nav_pos_switch) {
+
+        case USER_CONTROL_STABILISED:
+        {
+            //Operator is in control of roll and yaw.  Controls act as if in stabilise flight mode.  Pitch is controlled 
+            //by speed-height controller.
+            float pilot_roll, pilot_pitch;
+            get_pilot_desired_lean_angles(pilot_roll, pilot_pitch, copter.aparm.angle_max, copter.aparm.angle_max);
+
+            //get pilot's desired yaw rate
+            float pilot_yaw_rate = get_pilot_desired_yaw_rate(channel_yaw->get_control_in());
+
+            //pitch target is calculated in autorotation phase switch above
+            attitude_control->input_euler_angle_roll_pitch_euler_rate_yaw(pilot_roll, _pitch_target, pilot_yaw_rate);
+            //attitude_control->input_euler_angle_roll_pitch_euler_rate_yaw(pilot_roll, pilot_pitch, pilot_yaw_rate);
+            break;
+        }
+
+        case STRAIGHT_AHEAD:
```

Comment:
If these phases aren't used then I'd remove them or as a minimum compress them all together so it's like this:

```
case STRAIGHT_AHEAD:
case INTO_WIND:
case NEAREST_RALLY:
    break;
```

Reply:
compressed